src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs

```diff
@@ -415,6 +422,152 @@ private void HandleCodeActionRequest(string id, string paramsJson)
+
+            var tokenTypesToSkip = ParseTokenTypesToSkipParam(queryParams.Get("tokenTypesToSkip"));
+            var scope = _scopeFactory.GetOrCreateInstance(semanticTokensParams.TextDocument.Uri);
+            var checkResult = scope.Check(expression);
+
+            // Skip over the token types that clients don't want in the response
+            var tokens = checkResult.GetTokens().Where(tok => !tokenTypesToSkip.Contains(tok.TokenType));
+            var encodedTokens = SemanticTokensEncoder.EncodeTokens(tokens, expression, eol);
+
+            _sendToClient(JsonRpcHelper.CreateSuccessResult(id, new SemanticTokensResponse() { Data = encodedTokens }));
+        }
+
+        /// <summary>
+        /// Handles requests to compute semantic requests for the a specific part of document or expression.
+        /// </summary>
+        /// <param name="id">Request Id.</param>
+        /// <param name="paramsJson">Request Params Stringified Body.</param>
+        private void HandleRangeDocumentSemanticTokens(string id, string paramsJson)
+        {
+            if (!TryParseAndValidateSemanticTokenParams(id, paramsJson, out SemanticTokensRangeParams semanticTokensParams))
+            {
+                return;
+            }
+
+            if (semanticTokensParams.Range == null)
+            {
+                // No tokens for invalid range
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            var uri = new Uri(semanticTokensParams.TextDocument.Uri);
+            var queryParams = HttpUtility.ParseQueryString(uri.Query);
+            var expression = queryParams.Get("expression");
+
+            if (string.IsNullOrEmpty(expression))
+            {
+                // Empty tokens for the empty expression
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            // Monaco-Editor sometimes uses \r\n for the newline character. \n is not always the eol character so allowing clients to pass eol character
+            var eol = queryParams.Get("eol");
+            eol = !string.IsNullOrEmpty(eol) ? eol : EOL.ToString();
+
+            var (startIndex, endIndex) = PositionRangeHelper.ConvertRangeToPositions(semanticTokensParams.Range, expression, eol);
+            if (startIndex < 0 || endIndex < 0)
+            {
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            var tokenTypesToSkip = ParseTokenTypesToSkipParam(queryParams.Get("tokenTypesToSkip"));
+            var scope = _scopeFactory.GetOrCreateInstance(semanticTokensParams.TextDocument.Uri);
+            var result = scope.Check(expression);
+
+            // Skip over the token types that clients don't want in the response
+            var tokens = result.GetTokens().Where(tok => !tokenTypesToSkip.Contains(tok.TokenType));
+
+            // Only consider overlapping tokens. end index is exlcusive
```

Comment:
A bunch of this looks shared with HandleFullDocumentSemanticTokens, can some of it be factored out to a common function?

Reply:
Refactored